plugins/sonar-xoo-plugin/src/main/java/org/sonar/xoo/test/CoveragePerTestSensor.java

```diff
@@ -0,0 +1,122 @@
+import org.sonar.api.component.ResourcePerspectives;
+import org.sonar.api.test.MutableTestCase;
+import org.sonar.api.test.MutableTestPlan;
+import org.sonar.api.test.MutableTestable;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.xoo.Xoo;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Parse files *.xoo.testcoverage
+ */
+@DependsUpon("test-exec")
+public class CoveragePerTestSensor implements Sensor {
+  private static final Logger LOG = Loggers.get(CoveragePerTestSensor.class);
+
+  private static final String TEST_EXTENSION = ".testcoverage";
+
+  private final FileSystem fs;
+  private final ResourcePerspectives perspectives;
+
+  public CoveragePerTestSensor(FileSystem fileSystem, ResourcePerspectives perspectives) {
+    this.fs = fileSystem;
+    this.perspectives = perspectives;
+  }
+
+  private void processTestFile(InputFile inputFile, SensorContext context) {
+    File testExecutionFile = new File(inputFile.file().getParentFile(), inputFile.file().getName() + TEST_EXTENSION);
+    if (testExecutionFile.exists()) {
+      LOG.debug("Processing " + testExecutionFile.getAbsolutePath());
+      try {
+        List<String> lines = FileUtils.readLines(testExecutionFile, fs.encoding().name());
+        int lineNumber = 0;
+        MutableTestPlan testPlan = perspectives.as(MutableTestPlan.class, inputFile);
+        for (String line : lines) {
+          lineNumber++;
+          if (StringUtils.isBlank(line)) {
+            continue;
+          }
+          if (line.startsWith("#")) {
+            continue;
+          }
+          try {
+            Iterator<String> split = Splitter.on(";").split(line).iterator();
+            String name = split.next();
+            while (split.hasNext()) {
+              String coveredBlockStr = split.next();
+              Iterator<String> splitCoveredBlock = Splitter.on(",").split(coveredBlockStr).iterator();
+              String componentPath = splitCoveredBlock.next();
+              InputFile coveredFile = context.fileSystem().inputFile(context.fileSystem().predicates().hasPath(componentPath));
+              MutableTestable testable = perspectives.as(MutableTestable.class, coveredFile);
+              List<Integer> coveredLines = new ArrayList<>();
+              while (splitCoveredBlock.hasNext()) {
+                coveredLines.add(Integer.parseInt(splitCoveredBlock.next()));
+              }
+              for (MutableTestCase testCase : testPlan.testCasesByName(name)) {
+                testCase.setCoverageBlock(testable, coveredLines);
+              }
+            }
+          } catch (Exception e) {
+            throw new IllegalStateException("Error processing line " + lineNumber + " of file " + testExecutionFile.getAbsolutePath(), e);
+          }
+        }
+      } catch (IOException e) {
+        throw new RuntimeException(e);
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)Define and throw a dedicated exception instead of using a generic one.[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00112)